ctapipe/image/modifications.py

```diff
@@ -0,0 +1,231 @@
+        if to_smear == 0:
+            continue
+
+        # remove light from current pixel
+        new_image[pixel] -= to_smear
+
+        # add light to neighbor pixels
+        neighbors = indices[indptr[pixel] : indptr[pixel + 1]]
+        n_neighbors = len(neighbors)
+
+        # all neighbors are equally likely to receive the charge
+        # we always distribute the charge into 6 neighbors, so that charge
+        # on the edges of the camera is lost
+        neighbor_charges = np.random.multinomial(to_smear, smear_probabilities)
+
+        for n in range(n_neighbors):
+            neighbor = neighbors[n]
+            new_image[neighbor] += neighbor_charges[n]
+
+    return new_image
+
+
+def smear_psf_statically(image, geom, smear_factor):
+    """
+    Create a new image with values smeared to the direct pixel neighbors
+    Pixels at the camera edge lose charge this way.
+    This behaves differently for square and hexagonal pixels:
+    - For hexagonal pixels a fraction of light equal to smear_factor
+    in each pixel gets shifted to the neighboring pixels. Each pixel receives 1/6
+    of the light
+    - For square pixels, the image is converted to a 2d-array and a 3x3 gaussian
+    kernel is applied. Less light diffuses to diagonal neighbors. Smear factor
+    is the standard deviation of the gaussian kernel in this case
+
+    Parameters:
+    -----------
+    image: ndarray
+        1d array of the pixel charge values
+    geom: ctapipe.instrument.CameraGeometry
+    fraction: float
+        fraction of charge that will be distributed among neighbors
+
+    Returns:
+    --------
+    smeared_image: ndarray
+    """
+    if geom.pix_type is PixelShape.HEXAGON:
+        max_neighbors = 6
+        diffused_image = (
+            (image * geom.neighbor_matrix).sum(axis=1) * smear_factor / max_neighbors
+        )
+        remaining_image = image * (1 - smear_factor)
+        smeared_image = remaining_image + diffused_image
+    elif geom.pix_type is PixelShape.SQUARE:
+        image_2d = geom.to_regular_image(image)
+        # construct a normalized 3x3 kernel for convolution
+        kernel = np.zeros((3, 3))
+        kernel[1, 1] = 1
+        kernel = gaussian_filter(kernel, sigma=smear_factor)
+
+        smeared_2d = convolve2d(image_2d, kernel, mode="same")
+        smeared_image = geom.regular_image_to_1d(smeared_2d)
+    else:
+        raise Exception(f"Unknown pixel type {geom.pix_type}")
+    return smeared_image
+
+
+class ImageModifier(TelescopeComponent):
+    """
+    Abstract class for configurable image modifying algorithms. Use
+    ``ImageModifier.from_name()`` to construct an instance of a particular algorithm
+    """
+
+    @abstractmethod
+    def __call__(self, tel_id: int, image: np.ndarray) -> np.ndarray:
+        """
+        Modifies a given image
+
+        Returns
+        -------
+        np.ndarray
+            modified image
+        """
+
+
+class NullModifier(ImageModifier):
```

Comment:
Do we really need a "NullModifier"? Can't we just set the traitlet to `None` by default and do:

```
if self.image_transformer is not None:
    image = self.image_transformer(image)
```

Also: we might want to apply more than one, so is there an easy possibility to have multiple? Like

```
{
    "image_transformers": [
        {"cls": "NoiseAdder", ...},
        {"cls"}: "PSFSmearer", ...}
    ]
}
```


just thinking a bit

Reply:
1)
Probably not since we made the noise adder a noop by default anyway.

2)
I like it, but didn't we come to the conclusion that its not really feasible with the config/traitlets system currently? It would make a lot of sense if the order is fixed and classes can be used multiple times. Then you would be really flexible with your modifications. I don't remember exactly what the issue was though. Would this work on the command line?

Reply:
That is basically what is done with TelescopeParameters, but in this case, the problem is how to configure different instances of the same Component: the config system doesn't allow that currently.  Could be a future feature if we change config systems though!

Reply:
> the problem is how to configure different instances of the same Component
Thanks, that was probably it.

In that case I would propose to ignore the proposal for now because it leads to restrictions that are annoying and not super transparent to the user (mainly because the config files are not strictly checked before executing tools).
But your opinions may vary